### ISSUE TITLE
Use Result to further simplify client tasks and TUI actions

### ISSUE THREAD

**Issue by michaelsbradleyjr** [status-im/nim-status#287]
See: https://github.com/status-im/nim-status/pull/271#discussion_r690881289.

The existing pattern for *most* client tasks is pretty simple:
* Make a nim-status API call, passing through arguments (if any) received by the task.
* If the returned value `isErr` then send an event to the main thread (i.e. to the TUI) that has a non-empty error string.
* Otherwise send an event back to the main thread that has an empty error string and data derived from the returned value.

Some tasks do the above in multiple rounds.

The existing pattern for *most* TUI actions is also simple:
* If the event has a non-empty error string, print the error.
* Otherwise format and display the event's data.

So it seems like there's an opportunity to cut down on a lot of boilerplate given our design that nim-status API calls all return `Result[T, E]` or `Future[Result[T, E]]`.

Here's my initial proposal, minus exact details:

The base `Event` type (in `examples/client/client/common.nim`) would become `Result`.

In `examples/client/client/events.nim`, we'll have a table that maps nim-status API result types to client event types. 

There should be a `new` proc for each event type that accepts the corresponding API result and "translates" data/error into an event, which is itself a result. Each event type will need a corresponding type to hold translated data. For translated errors, `string` seems appropriate since they're destined for display by the TUI.

We'll have a macro, let's call it `wrap` for now, that will be used around API calls:

```nim
wrap(status.createAccount(12, "", password, dir))
```

`wrap` would make use of the table mapping:
```nim
const resultMappings = {
  "PublicAccountResult": "CreateAccountEvent"
}.toTable
```

And generate code like this:
```nim
let
  apiResult = status.createAccount(12, "", password, dir)
  event = CreateAccountEvent.new(apiResult)
  eventEnc = event.encode
  task = taskArg.taskName

if event.isErr:
  trace "task sent error event to host", error, task
  asyncSpawn chanSendToHost.send(eventEnc.safe)
  return

trace "task sent event to host", data=event.get, task
asyncSpawn chanSendToHost.send(eventEnc.safe)
```

`apiResult` and `event` would be made available in the scope where `wrap` was used.

`wrap` could take a few optional parameters: `returnOnErr = true, sendErr = true, sendOk = true`. Those could be used if we need to have code in the task body that will run after the code generated by `wrap()` *and* that code will be responsible for custom logic related to `chanSendToHost.send`.

Now, back on the main thread...

The behavior of [`dispatch`](https://github.com/status-im/nim-status/blob/master/examples/client/tui/events.nim#L11) would work a little differently than it does currently.

After decoding an event, if `isErr` then the error will be displayed in the TUI and `dispatch` will return, i.e. it won't proceed to call `action(decodedEvent)`.

Otherwise, it will call `action(decodedEvent.get)`.

So, all in all, this approach would cut down on much boilerplate in both client tasks and TUI actions.

**Comment by emizzle** [comment]
Firstly, great job opening this issue and kicking off the discussion!

It's important to point out a main differences between this suggestion above and the one proposed in https://github.com/status-im/nim-status/pull/271#discussion_r690881289. This solution attempts to allow sending of multiple events/errors to the host in each task, whereas the solution in #271 would send either an event *or* an error, but not multiple. I'd like to propose a solution that is similar to the solution above, but simplifies things a bit by not needing a mapping table and taking advantage of existing nim-result APIs.

> we'd have a table that maps nim-status API result types to client event types

As mentioned in https://github.com/status-im/nim-status/pull/271#discussion_r690859864, in #269 we have moved away from these specific Result types in favour of a generic pattern, for example going from:
```nim
# status/api/accounts.nim
type
  PublicAccountResult* = Result[PublicAccount, string]
```
to
```nim
# status/api/accounts.nim
type
  AccountsError* = enum
    ChatAcctNotFound = "accts: chat account not found in derived accounts"
    # ...

  AccountsResult*[T] = Result[T, AccountsError]
```
All modules in `status/api/*.nim` have been swapped over in favour of the generic pattern in #269 and so types like `PublicAccountResult` don't exist any longer. We should be considering the same pattern if considering a design change towards Results for the example client.

Due to the generic nature, I'm struggling to find a need for the mapping table and I'd much prefer to avoid yet another mapping table of stringified types if possible. The needed type information should be able to be made available to the compiler without a table.

To avoid a table, and keep it generic, we could alter the `wrap` usage a bit, use the existing nim-result APIs, and add our own convenience method, which should give the compiler all the information it needs. For example, we could go from the above suggestion:
```nim
# tasks.nim
wrap status.createAccount(12, "", password, dir)
```
to calling it `dispatch` instead of `wrap`, and using the available `nim-result` APIs to provide the compiler with what it needs:
```nim
# tasks.nim
dispatch status.createAccount(12, "", password, dir)
                  .flatMap(a => CreateAccountEvent(account: a)) # on ok, returns a `CreateAccountEvent` object
                  .mapErr(e => CreateAccountError) # on error, returns a TaskError enum value
```
This would then leave `dispatch` as a convenience method, without the need for complex macros:
```nim
proc dispatch*[V, E](self: Result[V, E]) =
  let
    eventEnc = (if self.isOk: self.get else: self.error).encode
    task = taskArg.taskName

  trace "task sent event to host", data=eventEnc, task
  asyncSpawn chanSendToHost.send(eventEnc.safe)
```
Note: it's likely `dispatch` may need to be a template or macro so that we can access context variables such as `task` and `chanSendToHost` without needing to pass them in explicitly.

`CreateAccountError` would actually be an error enum value, ie something like:
```nim
type
  TasksError* = enum
    CreateAccountError = "tasks: error creating account"
```

On the actions side, we'd be receiving a `TasksError` enum and it could be handled in the UI accordingly. This opens up possibilities for differentiating between different types of errors. We could, for example, have a `NetworkError` enum and display that in the "status bar" instead of in the main chat pane:

```nim
proc action*(self: Tui, event: CreateAccountEvent) {.async, gcsafe, nimcall.} =
  # display a successful message with the data from `CreateAccountEvent`
  let timestamp = getTime.toUnix
  self.printResult("Created account:", timestamp)
  self.printResult(fmt"{2.indent()}{name} ({abbrev})", timestamp)

proc action*(self: Tui, error: TasksError) {.async, gcsafe, nimcall.} =
  # display an error message relating to a TasksError
  # can distinguish between different task-related errors
  if error == CreateAccountError:
    self.wprintFormatError(getTime.toUnix, $error & ", too bad, so sad :(")

proc action*(self: Tui, error: NetworkError) {.async, gcsafe, nimcall.} =
  # display a network-related error
  # maybe here we want to display the result in the "status bar"
```

**Comment by michaelsbradleyjr** [comment]
Maybe it will be more clear by the time I've finished digesting what you've written (thanks for the detailed reply!), but I'm having trouble with this kind of thing:
```nim
type
  Foo = ...
  FooError = ...
  FooResult*[T] = Result[T, FooError]
```
***IF*** the only thing `T` will ever be is `Foo`.

In the case of `status/api/accounts.nim` I see *some* utility in `type AccountsResult*[T] = Result[T, AccountsError]` because `T` is sometimes `PublicAccount` and other times `Account`.

If `T` will only be one thing in practice, I can see preferring `Result[T, E]` in proc signatures as opposed to giving unique names to everything. What I mean is that instead of having `type FooResult[T] = Result[T, FooError]` or `type FooResult = [Foo, FooError]`, we could instead do this kind of thing:

```nim
# foo.nim
proc foo*(x: Result[Foo, FooError]): Result[Bar, BarError]
```
And the consumer of `foo.nim` can decide whether to *"give a name/alias to"* (create a type for) `Result[Foo, FooError]` or just write it out as `Result[...]`.

But then that circles back around to `type AccountsResult[T] = Result[T, AccountsError]`, which in proc signatures could be written instead as `Result[PublicAccount, AccountsError]` or `Result[Account, AccountsError]`, depending on the context.

It's not much more to type on the keyboard, we'd avoid a proliferation of type names, and my concern re: cases where *"`T` will only be one thing in practice"* simply melts away.

Anyway, rant over, I'll return to digesting your reply. 😃

**Comment by michaelsbradleyjr** [comment]
> I'd much prefer to avoid yet another mapping table of stringified types if possible

I figured out yesterday that with the help of some small macros we can easily create arrays or tables that are not stringified when we humans type them into the source code:
```nim
const tt = typesTable {
  Foo: Bar,
  Baz: Quux
}

const ta = typesArray [
  Foo,
  Bar,
  Baz,
  Quux
]
```

`typesTable` and `typesArray` are the macros we'd need to implement, but as mentioned, I already know how to do it.

**Comment by emizzle** [comment]
> In the case of `status/api/accounts.nim` I see *some* utility in `type AccountsResult*[T] = Result[T, AccountsError]` because `T` is sometimes `PublicAccount` and other times `Account`.

Yes, exactly, we are allowing for any return type, while keeping the error *at this layer of our abstraction* consistent. This allows higher layers to have a consistent error to make decisions upon receiving.

With that in mind, it doesn't buy us anything to keep the error type generic, as the case of `Result[T, E]`.

In regards to mapping tables, even if we did have a mapping table of types, we still would not be able to construct the mapped type unless we had some sort of inheritance structure with common fields, but looking at our currently implementation of Event types, I don't think this is the right way to go.
